diff --git a/ccache.c b/ccache.c
index 39742d9..290c688 100644
--- a/ccache.c
+++ b/ccache.c
@@ -501,8 +501,13 @@ static void from_cache(int first)
 
 	/* update timestamps for LRU cleanup
 	   also gives output_file a sensible mtime when hard-linking (for make) */
+#ifdef HAVE_UTIMES
+       utimes(hashname, NULL);
+	utimes(stderr_file, NULL);
+#else
 	utime(hashname, NULL);
 	utime(stderr_file, NULL);
+#endif
 
 	if (strcmp(output_file, "/dev/null") == 0) {
 		ret = 0;
diff --git a/ccache.h b/ccache.h
index 71b03ff..8d3c93c 100644
--- a/ccache.h
+++ b/ccache.h
@@ -22,6 +22,9 @@
 #ifdef HAVE_PWD_H
 #include <pwd.h>
 #endif
+#ifdef HAVE_SYS_TIME_H
+#include <sys/time.h>
+#endif
 
 #ifdef ENABLE_ZLIB
 #include <zlib.h>
diff --git a/config.h.in b/config.h.in
index 3068845..49ccd83 100644
--- a/config.h.in
+++ b/config.h.in
@@ -19,6 +19,9 @@
 /* Define to 1 if you have the `gethostname' function. */
 #undef HAVE_GETHOSTNAME
 
+/* Define to 1 if you have the `getpwuid' function. */
+#undef HAVE_GETPWUID
+
 /* Define to 1 if you have the <inttypes.h> header file. */
 #undef HAVE_INTTYPES_H
 
@@ -31,6 +34,9 @@
 /* Define to 1 if you have the <ndir.h> header file, and it defines `DIR'. */
 #undef HAVE_NDIR_H
 
+/* Define to 1 if you have the <pwd.h> header file. */
+#undef HAVE_PWD_H
+
 /* Define to 1 if you have the `realpath' function. */
 #undef HAVE_REALPATH
 
@@ -60,6 +66,9 @@
 /* Define to 1 if you have the <sys/stat.h> header file. */
 #undef HAVE_SYS_STAT_H
 
+/* Define to 1 if you have the <sys/time.h> header file. */
+#undef HAVE_SYS_TIME_H
+
 /* Define to 1 if you have the <sys/types.h> header file. */
 #undef HAVE_SYS_TYPES_H
 
@@ -69,6 +78,9 @@
 /* Define to 1 if you have the <unistd.h> header file. */
 #undef HAVE_UNISTD_H
 
+/* Define to 1 if you have the `utimes' function. */
+#undef HAVE_UTIMES
+
 /* Define to 1 if you have the `vasprintf' function. */
 #undef HAVE_VASPRINTF
 
diff --git a/configure.in b/configure.in
index be77bbf..1ac94d3 100644
--- a/configure.in
+++ b/configure.in
@@ -27,10 +27,11 @@ AC_HEADER_DIRENT
 AC_HEADER_TIME
 AC_HEADER_SYS_WAIT
 
-AC_CHECK_HEADERS(ctype.h strings.h stdlib.h string.h pwd.h)
+AC_CHECK_HEADERS(ctype.h strings.h stdlib.h string.h pwd.h sys/time.h)
 
 AC_CHECK_FUNCS(realpath snprintf vsnprintf vasprintf asprintf mkstemp)
 AC_CHECK_FUNCS(gethostname getpwuid)
+AC_CHECK_FUNCS(utimes)
 
 AC_CACHE_CHECK([for compar_fn_t in stdlib.h],ccache_cv_COMPAR_FN_T, [
     AC_TRY_COMPILE(
