diff --git a/ccache.c b/ccache.c
index 8ed9892..38775cf 100644
--- a/ccache.c
+++ b/ccache.c
@@ -144,7 +144,9 @@ static const char *tmp_string(void)
 		gethostname(hostname, sizeof(hostname)-1);
 #endif
 		hostname[sizeof(hostname)-1] = 0;
-		asprintf(&ret, "%s.%u", hostname, (unsigned)getpid());
+		if (asprintf(&ret, "%s.%u", hostname, (unsigned)getpid()) == -1) {
+			fatal("could not allocate tmp_string");
+		}
 	}
 
 	return ret;
diff --git a/stats.c b/stats.c
index ddaf768..593cd15 100644
--- a/stats.c
+++ b/stats.c
@@ -91,7 +91,7 @@ static void write_stats(int fd, unsigned counters[STATS_END])
 	if (len >= (int)sizeof(buf)-1) fatal("stats too long?!");
 
 	lseek(fd, 0, SEEK_SET);
-	write(fd, buf, len);
+	if (write(fd, buf, len) == -1) fatal("could not write stats");
 }
 
 
diff --git a/util.c b/util.c
index 6056d36..5b28d14 100644
--- a/util.c
+++ b/util.c
@@ -372,7 +372,9 @@ void x_asprintf(char **ptr, const char *format, ...)
 
 	*ptr = NULL;
 	va_start(ap, format);
-	vasprintf(ptr, format, ap);
+	if (vasprintf(ptr, format, ap) == -1) {
+		fatal("out of memory in x_asprintf");
+	}
 	va_end(ap);
 	
 	if (!ptr) fatal("out of memory in x_asprintf");
